src/emulate.c

```diff
@@ -1219,6 +1220,60 @@ RVOP(cswsp, {
 })
 #endif
 
+/* auipc + addi */
+RVOP(fuse1, {
+    rv->X[ir->rd] = (int32_t) (rv->PC + ir->imm + ir->imm2);
+    rv->PC += ir->insn_len;
+})
+
+/* auipc + add */
+RVOP(fuse2, {
+    rv->X[ir->rd] = (int32_t) (rv->X[ir->rs1]) + (int32_t) (rv->PC + ir->imm);
+    rv->PC += ir->insn_len;
+})
+
+/* multiple sw */
+RVOP(fuse3, {
+    opcode_fuse_t *fuse = ir->fuse;
+    uint32_t addr = rv->X[fuse[0].rs1] + fuse[0].imm;
+    /* the memory addresses of the sw instructions are contiguous, so we only
+     * need to check the first sw instruction to determine if its memory address
+     * is misaligned or if the memory chunk does not exist.
+     */
+    RV_EXC_MISALIGN_HANDLER(3, store, false, 1);
+    rv->io.mem_write_w(rv, addr, rv->X[fuse[0].rs2]);
+    for (int i = 1; i < ir->imm2; i++) {
+        addr = rv->X[fuse[i].rs1] + fuse[i].imm;
+        rv->io.mem_write_w(rv, addr, rv->X[fuse[i].rs2]);
+    }
+    rv->PC += ir->insn_len * (ir->imm2 - 1);
+})
+
+/* multiple lw */
```

Comment:
`lw` is the most frequent instruction (see #34), and we might dive into its use case more.

Reply:
Can you handle the following case? (disassembly from CoreMark
```
   10248:       03012603                lw      a2,48(sp)
   1024c:       01c11583                lh      a1,28(sp)
   10250:       03412503                lw      a0,52(sp)
```

Reply:
In addition, consider the following scenario:
```
   10a84:       01c12083                lw      ra,28(sp)
   10a88:       07f47513                andi    a0,s0,127
   10a8c:       01812403                lw      s0,24(sp)
   10a90:       01412483                lw      s1,20(sp)
   10a94:       01012903                lw      s2,16(sp)
   10a98:       00c12983                lw      s3,12(sp)
```

It can be regarded as 5 `lw`. Roughly speaking, if [peephole optimization](https://en.wikipedia.org/wiki/Peephole_optimization) can be applied, we shall benefit from further optimizations.

Reply:
Another case: (disassembly from CoreMark)
```
   10c08:       01162023                sw      a7,0(a2)
   10c0c:       00052783                lw      a5,0(a0)
   10c10:       00059883                lh      a7,0(a1)
   10c14:       00259603                lh      a2,2(a1)
   10c18:       00f82023                sw      a5,0(a6)
   10c1c:       01052023                sw      a6,0(a0)
   10c20:       00e82223                sw      a4,4(a6)
   10c24:       0006a783                lw      a5,0(a3)
```

Mixture of `sw` and `lw`.

Reply:
> Another case: (disassembly from CoreMark)
> 
> ```
>    10c08:       01162023                sw      a7,0(a2)
>    10c0c:       00052783                lw      a5,0(a0)
>    10c10:       00059883                lh      a7,0(a1)
>    10c14:       00259603                lh      a2,2(a1)
>    10c18:       00f82023                sw      a5,0(a6)
>    10c1c:       01052023                sw      a6,0(a0)
>    10c20:       00e82223                sw      a4,4(a6)
>    10c24:       0006a783                lw      a5,0(a3)
> ```
> 
> Mixture of `sw` and `lw`.

In this case, the memory address is not contiguous, what we can do just pack these instructions, but we cannot save any operation, such as checking misaligned.

Reply:
> In addition, consider the following scenario:
> 
> ```
>    10a84:       01c12083                lw      ra,28(sp)
>    10a88:       07f47513                andi    a0,s0,127
>    10a8c:       01812403                lw      s0,24(sp)
>    10a90:       01412483                lw      s1,20(sp)
>    10a94:       01012903                lw      s2,16(sp)
>    10a98:       00c12983                lw      s3,12(sp)
> ```
> 
> It can be regarded as 5 `lw`. Roughly speaking, if [peephole optimization](https://en.wikipedia.org/wiki/Peephole_optimization) can be applied, we shall benefit from further optimizations.

In this case, we can pack the last four instruction lw. if we want to handle this case by packing 5 lw, we need to reorder the instruction. For example, swap the first and the second instruction.

Reply:
> Can you handle the following case? (disassembly from CoreMark
> 
> ```
>    10248:       03012603                lw      a2,48(sp)
>    1024c:       01c11583                lh      a1,28(sp)
>    10250:       03412503                lw      a0,52(sp)
> ```

Ditto, if we want to handle this case, we need some strategies to reorder the instructions.

Reply:
In this pull request, let's concentrate on preliminary support of macro operation fusion. You shall add some comments for further efforts such as instruction reordering.